src/splay_tree.c

```diff
@@ -604,12 +604,22 @@ void splay_delete(splay_tree_t *tree, void *data) {
 
 /* Fast tree cleanup */
 
-void splay_delete_tree(splay_tree_t *tree) {
+void splay_free_nodes(splay_tree_t *tree) {
 	for(splay_node_t *node = tree->head, *next; node; node = next) {
 		next = node->next;
 		splay_free_node(tree, node);
 	}
+}
```

Comment:
This is a potential landmine; it doesn't zero `tree->head` and related pointers. I also don't see why we need both `splay_free_nodes()` and the slower `splay_delete_nodes()`. If we are going to remove all the nodes we don't need to unlink them first. I think it's better to create one function named `splay_empty_tree()`, like in @fangfufu's patch. Call `splay_free_nodes()` and then set `tree->head`, `tree->tail` and `tree->root` all to `NULL`.

Reply:
Also `tree->count = 0`, and `tree->generation++`.

Reply:
That's much better. Thank you.

I am not sure if it makes sense to inline the remaining two configuration trees in `connection_t` and `address_cache_t`. They are lazily initialized and (I think) can be reused many times over. Initialization relies on some data that is not known when the structure itself is allocated (like node name). This can be circumvented with additional flags, etc, but it seems to only make code worse than it is now.

https://github.com/gsliepen/tinc/blob/0871c3095151bce6a4031a2662aa51b7193b855c/src/connection.h#L107-L108

https://github.com/gsliepen/tinc/blob/95230693d2fda7bde6d44aea170ae12caf8eab7a/src/address_cache.h#L28-L30